services/src/pro/quota/mod.rs

```diff
@@ -0,0 +1,161 @@
+use futures::{
+    channel::mpsc::{channel, Receiver, Sender},
+    StreamExt,
+};
+use geoengine_operators::pro::meta::quota::{ComputationContext, ComputationUnit, QuotaTracking};
+use std::sync::Arc;
+
+use geoengine_datatypes::util::test::TestDefault;
+
+use crate::pro::users::UserId;
+
+use super::users::{UserDb, UserSession};
+
+// TODO: find a good default, choose w.r.t. resources or make configurable
+const QUOTA_CHANNEL_SIZE: usize = 1000;
```

Comment:
What happens if this channel is bounded? The Futures were spawned in the Stream adapter… and… then?

I'm not sure what makes sense here.

Reply:
> Do we want to block our computations when our database or our implementation of logging is not fast enough?

right now, we wouldn't. Only the task we spawn would block. Computation continues.

> Do we run into the issue of having 4000 futures that use memory and block because the channel is full?

Thinking of all the stuff that is done in order to produce a chunk, I doubt that we run into the problem of having to many quota futures.

> Is it more reasonable to limit the channel to some number or to have it unlimited if we have no strategy to cope with resolving it?

I thought an unbounded channel had some limitations? At least the std::sync::Sender was not Sync but the SyncSender was and that one was bounded. If we can interchange the tokio channel with an unbounded one I would say we should do it.

Reply:
1. You are right that the future just spawns into the wild!

2. 
https://docs.rs/tokio/latest/tokio/sync/mpsc/index.html

It has 
```
[Functions](https://docs.rs/tokio/latest/tokio/sync/mpsc/index.html#functions)
[channel](https://docs.rs/tokio/latest/tokio/sync/mpsc/fn.channel.html)
Creates a bounded mpsc channel for communicating between asynchronous tasks with backpressure.
[unbounded_channel](https://docs.rs/tokio/latest/tokio/sync/mpsc/fn.unbounded_channel.html)
Creates an unbounded mpsc channel for communicating between asynchronous tasks without backpressure.
```

I'm not entirely against a bounded channel, I'm just curious what the implications might be.

Reply:
I think there are no implications because updating the quota in the database will always be faster than creating new tiles. So the channel should never fill up.

Reply:
Okay, so… make it unbounded then?

Btw: I'm not 100% sure that creating a tile is always slower then accessing an external database, starting a transaction, updating a record and submitting the transaction.

Reply:
> Btw: I'm not 100% sure that creating a tile is always slower then accessing an external database, starting a transaction, updating a record and submitting the transaction.

it doesn't have to be always slower, just on average, so that the bottleneck is actually creating the tiles and not updating the quota in the database. I changed it to an unbounded channel now, so now the bottleneck is our ram which may be filled up with pending quota messages that are not yet written to the database.

3715bcf54d31857d07e77e3b748076cf72108b6f